ip_test.go

```diff
@@ -1,235 +1,605 @@
-	ipTestReqKeyRealIPAndXFFExternal = "x-real-ip and xff; remote addr external"
-	ipTestReqKeyRealIPAndXFFInternal = "x-real-ip and xff; remote addr internal"
-	ipTestReqKeyXFFExternal          = "xff; remote addr external"
-	ipTestReqKeyXFFInternal          = "xff; remote addr internal"
-	ipTestReqKeyBrokenXFF            = "broken xff"
-)
+func mustParseCIDR(s string) *net.IPNet {
+	_, IPNet, err := net.ParseCIDR(s)
+	if err != nil {
+		panic(err)
+	}
+	return IPNet
+}
+
+func TestIPChecker_TrustOption(t *testing.T) {
+	var testCases = []struct {
+		name         string
+		givenOptions []TrustOption
+		whenIP       string
+		expect       bool
+	}{
+		{
+			name: "ip is within trust range, trusts additional private IPV6 network",
+			givenOptions: []TrustOption{
+				TrustLoopback(false),
+				TrustLinkLocal(false),
+				TrustPrivateNet(false),
+				// this is private IPv6 ip
+				// CIDR Notation: 	2001:0db8:0000:0000:0000:0000:0000:0000/48
+				// Address: 				2001:0db8:0000:0000:0000:0000:0000:0103
+				// Range start: 		2001:0db8:0000:0000:0000:0000:0000:0000
+				// Range end: 			2001:0db8:0000:ffff:ffff:ffff:ffff:ffff
+				TrustIPRange(mustParseCIDR("2001:db8::103/48")),
+			},
+			whenIP: "2001:0db8:0000:0000:0000:0000:0000:0103",
+			expect: true,
+		},
+		{
+			name: "ip is within trust range, trusts additional private IPV6 network",
+			givenOptions: []TrustOption{
+				TrustIPRange(mustParseCIDR("2001:db8::103/48")),
+			},
+			whenIP: "2001:0db8:0000:0000:0000:0000:0000:0103",
+			expect: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			checker := newIPChecker(tc.givenOptions)
+
+			result := checker.trust(net.ParseIP(tc.whenIP))
+			assert.Equal(t, tc.expect, result)
+		})
+	}
+}
+
+func TestTrustIPRange(t *testing.T) {
+	var testCases = []struct {
+		name       string
+		givenRange string
+		whenIP     string
+		expect     bool
+	}{
+		{
+			name: "ip is within trust range, IPV6 network range",
+			// CIDR Notation: 	2001:0db8:0000:0000:0000:0000:0000:0000/48
+			// Address: 				2001:0db8:0000:0000:0000:0000:0000:0103
+			// Range start: 		2001:0db8:0000:0000:0000:0000:0000:0000
+			// Range end: 			2001:0db8:0000:ffff:ffff:ffff:ffff:ffff
+			givenRange: "2001:db8::103/48",
+			whenIP:     "2001:0db8:0000:0000:0000:0000:0000:0103",
+			expect:     true,
+		},
+		{
+			name:       "ip is outside (upper bounds) of trust range, IPV6 network range",
+			givenRange: "2001:db8::103/48",
+			whenIP:     "2001:0db8:0001:0000:0000:0000:0000:0000",
+			expect:     false,
+		},
+		{
+			name:       "ip is outside (lower bounds) of trust range, IPV6 network range",
+			givenRange: "2001:db8::103/48",
+			whenIP:     "2001:0db7:ffff:ffff:ffff:ffff:ffff:ffff",
+			expect:     false,
+		},
+		{
+			name: "ip is within trust range, IPV4 network range",
+			// CIDR Notation: 	8.8.8.8/24
+			// Address: 				8.8.8.8
+			// Range start: 		8.8.8.0
+			// Range end: 			8.8.8.255
+			givenRange: "8.8.8.8/24",
```

Comment:
Usually a trust subnet would be specified using the base address with subnet mask, in that case 8.8.8.0/24. Unless this was intentional, maybe we should change that.
(also not really important)

Reply:
seems that Echo `.editorconfig` sets `indent_size = 2`